Source/Core/Core/PowerPC/Interpreter/Interpreter_LoadStorePaired.cpp

```diff
@@ -48,212 +48,269 @@ const float m_quantizeTable[] =
+		}
+		else
 		{
-			float fResult = (float)_fValue * m_quantizeTable[_uScale];
-			MathUtil::Clamp(&fResult, -128.0f, 127.0f);
-			PowerPC::Write_U8((u8)(s8)fResult, _Addr);
+			u8 convPS0 = (u8)ScaleAndClamp<s8>(ps0, stScale);
+			u8 convPS1 = (u8)ScaleAndClamp<s8>(ps1, stScale);
+			WritePair8(convPS0, convPS1, addr);
 		}
 		break;
 
 	case QUANTIZE_S16:
+		if (instW)
 		{
-			float fResult = (float)_fValue * m_quantizeTable[_uScale];
-			MathUtil::Clamp(&fResult, -32768.0f, 32767.0f);
-			PowerPC::Write_U16((u16)(s16)fResult, _Addr);
+			u16 convPS0 = (u16)ScaleAndClamp<s16>(ps0, stScale);
+			PowerPC::Write_U16(convPS0, addr);
+		}
+		else
+		{
+			u16 convPS0 = (u16)ScaleAndClamp<s16>(ps0, stScale);
+			u16 convPS1 = (u16)ScaleAndClamp<s16>(ps1, stScale);
+			WritePair16(convPS0, convPS1, addr);
 		}
 		break;
 
-	default:
-		_dbg_assert_msg_(POWERPC, 0, "PS dequantize - unknown type to read");
+	case QUANTIZE_INVALID1:
+	case QUANTIZE_INVALID2:
+	case QUANTIZE_INVALID3:
+		_assert_msg_(POWERPC, 0, "PS dequantize - unknown type to read");
 		break;
 	}
 }
 
-float Interpreter::Helper_Dequantize(const u32 _Addr, const EQuantizeType _quantizeType, const unsigned int _uScale)
+void Interpreter::Helper_Dequantize(u32 addr, u32 instI, u32 instRD, u32 instW)
 {
-	// dequantize the value
-	float fResult;
-	switch (_quantizeType)
+	UGQR gqr(rSPR(SPR_GQR0 + instI));
+	EQuantizeType ldType = gqr.ld_type;
+	unsigned int ldScale = gqr.ld_scale;
+
+	float ps0, ps1;
+	switch (ldType)
 	{
 	case QUANTIZE_FLOAT:
+		if (instW)
+		{
+			u32 value = PowerPC::Read_U32(addr);
+			ps0 = MathUtil::IntFloat(value).f;
+			ps1 = 1.0f;
+		}
+		else
 		{
-			u32 dwValue = PowerPC::Read_U32(_Addr);
-			fResult = *(float*)&dwValue;
+			std::pair<u32, u32> value = ReadPair32(addr);
+			ps0 = MathUtil::IntFloat(value.first).f;
+			ps1 = MathUtil::IntFloat(value.second).f;
 		}
 		break;
 
 	case QUANTIZE_U8:
-		fResult = static_cast<float>(PowerPC::Read_U8(_Addr)) * m_dequantizeTable[_uScale];
+		if (instW)
+		{
+			u8 value = PowerPC::Read_U8(addr);
+			ps0 = (float)(u8)(value) * m_dequantizeTable[ldScale];
+			ps1 = 1.0f;
+		}
+		else
+		{
+			std::pair<u8, u8> value = ReadPair8(addr);
+			ps0 = (float)(u8)(value.first) * m_dequantizeTable[ldScale];
+			ps1 = (float)(u8)(value.second) * m_dequantizeTable[ldScale];
+		}
 		break;
 
 	case QUANTIZE_U16:
-		fResult = static_cast<float>(PowerPC::Read_U16(_Addr)) * m_dequantizeTable[_uScale];
+		if (instW)
+		{
+			u16 value = PowerPC::Read_U16(addr);
+			ps0 = (float)(u16)(value) * m_dequantizeTable[ldScale];
+			ps1 = 1.0f;
+		}
+		else
+		{
+			std::pair<u16, u16> value = ReadPair16(addr);
+			ps0 = (float)(u16)(value.first) * m_dequantizeTable[ldScale];
+			ps1 = (float)(u16)(value.second) * m_dequantizeTable[ldScale];
+		}
 		break;
 
 	case QUANTIZE_S8:
-		fResult = static_cast<float>((s8)PowerPC::Read_U8(_Addr)) * m_dequantizeTable[_uScale];
+		if (instW)
+		{
+			u8 value = PowerPC::Read_U8(addr);
```

Comment:
If PowerPC::Read_xxx, PowerPC::Write_xxx_Swap, and Common::swapXX used templates you could make ReadPairXX templated, and then generalize all of these into a templated function as well, returning ps0 and ps1 as a pair. The changes would probably stem beyond the scope of this PR though.